### DISCUSSION TITLE
Error Occurs with revalidateTag and header() Usage in Middleware

### DISCUSSION THREAD

**Discussion by hwangyena** [vercel/next.js#63395] [Help]
### Summary

When using `fetch` and `header()` in middleware, and using revalidateTag or revalidatePath, a 500 error occurs.
The code was used as follows.

middleware.ts
```typescript
export const middleware = async (request:NextRequest) => {
   const res = await fetch(`http://${request.headers.get('host')}/api/session`, { headers: request.headers }); // ERROR!
   ...
```

api.ts
```typescript
export async function createComment({ id, content }: { id: string; content: string }) {
  try {
    return await fetch(`${process.env.NEXT_PUBLIC_VERCEL_URL}/api/post/${id}/comment`, {
      method: 'POST',
      body: JSON.stringify({ content }),
    })
  } ...
}
```

revalidate.ts
```typescript
'use server';

import { revalidateTag } from 'next/cache';

export async function revalidateFetchTag(tag: string) {
  revalidateTag(tag);
}
```

component.tsx
```typescript
const onCreateComment = useCallback(async () => {
    await createComment({ id: postId, content });

    revalidateFetchTag('comment');
}, [content, postId, router]);
```

Here is the error message. 
<img width="557" alt="image" src="https://github.com/vercel/next.js/assets/70925962/688f22e5-68a5-45a6-b683-95bd964eb91c">

When I tried using both revalidateTag and revalidatePath, I encountered the same error. 
If I don't request request.headers in the middleware, the error doesn't occur. What could be the reason for this error?

**Comment by icyJoseph** [comment]
I can speculate that it is not a good idea to do this:

```js
fetch(`http://${request.headers.get('host')}/api/session`, { headers: request.headers })
```

That is, to pick the `host` out of the request and then try to make an absolute URL with that. Why are you even doing that? Middleware is already running server side, so you can do, for the most part, whatever you have in `api/session` already here in the middleware.

You are unnecessarily making all of the requests to your site slower, and dependent on a round trip to the same server. Are you having issues with certain Node.js modules not being available in Middleware? There's way to fix that instead.

**Reply by hwangyena** [reply]
Thank you for your response.
We are currently using next-auth in the project, but we're not using JWT, so we used this code. 
It turns out that using `getServerSideSession` directly in the middleware causes an error. 
Is there any other alternative solution to this problem?

**Reply by kirim-Lee** [reply]
when use `getServerSession` at middleware in next-auth ver.4, it occures error 
so it is a detour by call absolute path directly and it's route returns session using getServerSession.
using getSession can avoid this solution, but official document show getSession is client method
https://next-auth.js.org/tutorials/securing-pages-and-api-routes
https://github.com/nextauthjs/next-auth/issues/7732
https://github.com/nextauthjs/next-auth/discussions/4265

**Reply by kirim-Lee** [reply]
if it will occur same error when we have neccesity to use outer api in middleware, use req.headers should be used, so the error must be solved

**Reply by icyJoseph** [reply]
Try with `headers: new Headers(request.headers)`

**Reply by hwangyena** [reply]
I've also tried using `new Header`, but it still doesn't resolve the issue. 🥲
